velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -17,15 +17,136 @@
+            TEMPLATE_FUNC, typeKind, __VA_ARGS__);                        \
+      }                                                                   \
+      default:                                                            \
+        VELOX_UNREACHABLE(                                                \
+            "Unsupported partition type: ", mapTypeKindToName(typeKind)); \
+    }                                                                     \
+  }()
+
+namespace {
+std::vector<column_index_t> getPartitionChannels(
+    const std::shared_ptr<const HiveInsertTableHandle>& insertTableHandle) {
+  std::vector<column_index_t> channels;
+
+  for (column_index_t i = 0; i < insertTableHandle->inputColumns().size();
+       i++) {
+    if (insertTableHandle->inputColumns()[i]->isPartitionKey()) {
+      channels.push_back(i);
+    }
+  }
+
+  return channels;
+}
+
+// Compute the count of rows as well as the concrete row indices for every
+// partition ID based on the ID labeling of rowPartitionIds. Both rowCounts
+// and rowIndices are indexed by partition ID. rowIndices[i] is a vector of
+// all rows labeled with partition ID i.
+void computePartitionRowCountsAndIndices(
+    const raw_vector<uint64_t>& partitionIds,
+    uint64_t numPartitionIds,
+    memory::MemoryPool* FOLLY_NONNULL pool,
+    std::vector<vector_size_t>& partitionSizes,
+    std::vector<BufferPtr>& partitionRows) {
+  partitionSizes.resize(numPartitionIds);
+  std::fill(partitionSizes.begin(), partitionSizes.end(), 0);
+  partitionRows.resize(numPartitionIds);
+  std::fill(partitionRows.begin(), partitionRows.end(), nullptr);
+
+  std::vector<vector_size_t*> rawIndexValues(numPartitionIds);
+  uint64_t numRows = partitionIds.size();
+
+  for (vector_size_t row = 0; row < numRows; row++) {
+    uint64_t id = partitionIds[row];
+
+    if (partitionRows[id] == nullptr) {
+      partitionRows[id] = allocateIndices(numRows, pool);
+      rawIndexValues[id] = partitionRows[id]->asMutable<vector_size_t>();
+    }
+
+    rawIndexValues[id][partitionSizes[id]] = row;
+    partitionSizes[id]++;
+  }
+
+  for (auto id = 0; id < numPartitionIds; id++) {
+    if (partitionRows[id] != nullptr) {
+      partitionRows[id]->setSize(partitionSizes[id] * sizeof(vector_size_t));
+    }
+  }
+}
+
+// TODO(gaoge): escape path characters as in
+// https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/FileUtils.java
+template <typename T>
+inline std::string makeColumnString(T value) {
+  return folly::to<std::string>(value);
+}
+
+template <>
+inline std::string makeColumnString(Date value) {
+  return value.toString();
+}
+
+template <TypeKind Kind>
+std::string makePartitionColumnString(
+    const BaseVector* partitionVector,
+    vector_size_t row,
+    const std::string& name) {
+  using T = typename TypeTraits<Kind>::NativeType;
+  auto* simpleVector = partitionVector->as<SimpleVector<T>>();
+  VELOX_CHECK_NOT_NULL(simpleVector, "");
```

Comment:
Error message is empty string. Is this intentional?

Reply:
Nope :(